Add support for lucide icons

kocheick opened the pull request:
## Overview

Implements support for [Lucide Icons](https://lucide.dev/icons/) in Kobweb with build-time code generation and
tree-shaking optimization, addressing issue #697.

## What’s Changed

### New Module: `silk-icons-lucide`

- Created new frontend module with npm dependency on Lucide icons
- Implemented build-time codegen that extracts SVG data from Lucide’s ESM modules
- Generated strongly-typed composable functions for all 1,600+ Lucide icons
- Added tree-shaking support — only used icons are included in the final bundle

### API Design

```kotlin
// Default usage
LiCamera()

// Customized with size and color
LiHeart(size = 2.em, color = Colors.Red)
```

- Icon functions use the Li prefix (Lucide Icons). Example mappings:
    - camera → LiCamera
    - arrow-down → LiArrowDown
- One composable per icon for optimal tree-shaking.
- No dynamic/generic fallback API in this iteration.

## Key Features

- **Tree-shakable**: Only imported icons are included in the final bundle
- **Type-safe**: Individual composables for each icon (e.g., `LiCamera`, `LiHeart`)
- **Customizable**: Size, stroke width, color, and standard modifiers
- **Performance**: Renders as inline SVG elements for optimal performance
- **Build-time generation**: Icons are generated during build from the npm package

## Usage Examples

- Default icon usage: `LiCamera()`
- Customized icon: `LiZoomIn(size = 2.em, color = Colors.Honeydew)`

## Documentation

- Updated module README with Li* usage examples and API reference
- Clear examples showing default and customized icon usage

## Files Added/Modified

- `frontend/silk-icons-lucide/build.gradle.kts` — Build configuration with codegen
- `frontend/silk-icons-lucide/README.md` — Usage documentation (updated to Li*)
- `playground/site/src/jsMain/kotlin/playground/pages/Widgets.kt` — Usage examples (updated to Li*)
- Generated: `LucideIcons.kt` with all Li* icon composables

## Testing

- Icons render correctly in dev and production builds
- Tree-shaking intent: only used icons are included (to be verified)
- SVG structure matches Lucide specifications
- Custom styling (size, color, stroke-width) works as expected

## Breaking Changes

None — this is a new optional module.

## Related Issues

Closes #697

kocheick committed:
Add Lucide icon support to the project

kocheick committed:
Include Lucide icons in aggregate-docs build configuration

kocheick committed:
Add Lucide icon support to playground site build configuration

kocheick committed:
add lucide icon support to build configuration and include in aggregate docs

kocheick committed:
Add example usage of Lucide icons to playground site index page

kocheick committed:
Update Lucide icon examples on the playground site index page

bitspittle commented on frontend/silk-icons-lucide/build.gradle.kts:
```diff
@@ -0,0 +1,322 @@
+    
+    outputs.dir("node_modules")
+    
+    doFirst {
```
I'd prefer just copying the json file in manually and including a README with instructions on how to regenerate it. I'm very nervous about adding any build step that requires connecting to the internet.

bitspittle commented on playground/site/src/jsMain/kotlin/playground/pages/Index.kt:
Don't modify Index.kt but instead edit Widgets.kt. In fact, you'll see that there is already a section for font awesome and mdi icons in there.

bitspittle commented on frontend/silk-icons-lucide/build.gradle.kts:
```diff
@@ -0,0 +1,322 @@
+    color: CSSColorValue? = null
+) {
+    when (name) {
+        $iconCases
```
I'm pretty sure this approach prevents dead code elimination from working. If I use just a single Lucide icon, I'm going to be referencing a function that is going to add a TON of size to the final JS output.

So, in short, let's not provide a raw LucideIcon composable. Instead, just embed render logic into the lucide icons body directly, e.g.:

```kotlin
@Composable
fun LucideAArrowDown(
    modifier: Modifier = Modifier,
    size: CSSLengthValue = 1.em,
    strokeWidth: Number = 2,
    color: CSSColorValue? = null
) {
   renderIcon(listOf("path" to mapOf("d" to "m14 12 4 4 4-4"), "path" to mapOf("d" to "M18 16V7"), "path" to mapOf("d" to "m2 16 4.039-9.69a.5.5 0 0 1 .923 0L11 16"), "path" to mapOf("d" to "M3.304 13h6.392")), modifier, size, strokeWidth, color)
}
```

kocheick replied:
will refactor these asap.
By the way, i was thinking of renaming the generated function names to LuCamera() instead of LucideCamera(), LuAArrowDown, etc.. just to follow Fa and Mdi conventions

bitspittle replied:
Sounds good! Although maybe Li for Lucide Icons, the same way Mdi is Material Design Icons?

kocheick replied:
Good, Li for Lucide Icons it is. now upating ths PR docs and the README file

bitspittle reviewed (changes requested):
Please check the README.mds I added in silk-icons-fa and silk-icons-mdi. It would be nice if you could add one for this feature as well. It shouldn't be long -- just add some basic information about where you get lucide icons data from, things that might help a future dev navigate this module if they need to make changes.

kocheick committed:
Remove Lucide icon examples from the playground site index page

kocheick committed:
Simplify Lucide icon generation by removing npm dependency and using pre-parsed JSON data

kocheick committed:
Simplify Lucide icon generation by removing npm dependency and using pre-parsed JSON data

kocheick committed:
Add pre-parsed Lucide icon JSON data to the project for easier usage

kocheick committed:
Add pre-parsed Lucide icon JSON data to the project for easier usage

kocheick committed:
Revert "Add pre-parsed Lucide icon JSON data to the project for easier usage"

This reverts commit a672d28099420008b01710cb15bf4c75ab19d53f.

kocheick committed:
update README documentation for Silk Icons (Lucide) detailing usage, setup, and features.

kocheick committed:
Add error handling for missing lucide-icons.json with detailed instructions

kocheick committed:
Rename "Lucide" references to "Lu" in Widgets.kt for consistency

kocheick committed:
Add script to extract Lucide icon data and generate `lucide-icons.json`

kocheick committed:
rename lucide icon references to li for consistency and update related documentation and build scripts accordingly

kocheick committed:
Merge branch 'dev' into Add-support-for-Lucide-Icons

kocheick committed:
update README to clarify lucide icon naming and include recent file and usage updates

kocheick committed:
Merge remote-tracking branch 'origin/Add-support-for-Lucide-Icons' into Add-support-for-Lucide-Icons

kocheick committed:
update README to specify icon imports explicitly for clarity

kocheick committed:
removed "  |//@formatter:off
                     |@file:Suppress("unused", "SpellCheckingInspection")
                     |"

                     from generated LucideIcons.kt
